src/main/java/sirius/db/es/ElasticQuery.java

```diff
@@ -1055,8 +1052,25 @@ protected void doIterate(Predicate<E> handler) {
                                       limit,
                                       buildPayload());
         for (JsonNode obj : Json.getArrayAt(this.response, HITS_POINTER)) {
-            if (!handler.test((E) Elastic.make(descriptor, (ObjectNode) obj))) {
-                return;
+            // This is the mist common use case, so we handle it first...
+            if (additionalDescriptors == null || additionalDescriptors.isEmpty()) {
+                ElasticEntity entity = Elastic.make(descriptor, (ObjectNode) obj);
+                if (!handler.test((E) entity)) {
+                    return;
+                }
+            } else {
+                String indexName = obj.get("_index").asText(null);
+                ElasticEntity entity = additionalDescriptors.stream()
+                                                            .filter(additionalDescriptor -> additionalDescriptor.getRelationName()
+                                                                                                                .equals(indexName))
+                                                            .findFirst()
+                                                            .map(matchingDescriptor -> Elastic.make(matchingDescriptor,
+                                                                                                    (ObjectNode) obj))
+                                                            .orElseGet(() -> Elastic.make(descriptor,
+                                                                                          (ObjectNode) obj));
+                if (!handler.test((E) entity)) {
+                    return;
+                }
```

Comment:
(optional) maybe extract a method here ElasticEntity entity = extractEntity(...) - and have only one if (!handler.test((E) entity)) { as remaining call - and also a shorter for loop